k9mail/src/main/java/com/fsck/k9/activity/setup/AccountSetupPresenter.java

```diff
@@ -243,14 +280,89 @@ protected ProviderInfo doInBackground(Void... params) {
+                if (providerInfo != null) return providerInfo;
 
-                return providerInfo;
+                testDomain(domain);
+
+                return null;
+            }
+
+            private void testDomain(String domain) {
+                publishProgress(R.string.account_setup_check_settings_guessing_msg);
+
+                String guessedDomainForMailPrefix;
+                //noinspection ConstantConditions
+                if (domain.startsWith("mail.")) {
+                    guessedDomainForMailPrefix = domain;
+                } else {
+                    guessedDomainForMailPrefix = "mail." + domain;
+                }
+
+                testIncoming(guessedDomainForMailPrefix);
+
+                testOutgoing(guessedDomainForMailPrefix, ConnectionSecurity.STARTTLS_REQUIRED);
+
+                testOutgoing(guessedDomainForMailPrefix, ConnectionSecurity.SSL_TLS_REQUIRED);
+
+                testIncoming("imap." + domain);
+
+                testOutgoing("smtp." + domain, ConnectionSecurity.STARTTLS_REQUIRED);
+
+                testOutgoing("smtp." + domain, ConnectionSecurity.SSL_TLS_REQUIRED);
+            }
+
+            private void testIncoming(String domain) {
+                if (!incomingReady) {
+                    try {
+                        accountConfig.setStoreUri(getDefaultStoreURI(email, password, domain).toString());
+                        accountConfig.getRemoteStore().checkSettings();
+                        incomingReady = true;
+                    } catch (URISyntaxException | MessagingException ignored) {
+                    }
+                }
+            }
+
+            private void testOutgoing(String domain, ConnectionSecurity connectionSecurity) {
+                if (!outgoingReady) {
+                    try {
+                        accountConfig.setTransportUri(getDefaultTransportURI(email, password, domain,
+                                connectionSecurity).toString());
+                        Transport transport = TransportProvider.getInstance().getTransport(context, accountConfig,
+                                Globals.getOAuth2TokenProvider());
+                        transport.close();
+                        try {
+                            transport.open();
+                        } finally {
+                            transport.close();
+                        }
+                        outgoingReady = true;
+                    } catch (URISyntaxException | MessagingException ignored) {
+                    }
+                }
+            }
+
+            @Override
+            protected void onProgressUpdate(Integer... values) {
+                super.onProgressUpdate(values);
+                view.setMessage(values[0]);
             }
 
             @Override
             protected void onPostExecute(ProviderInfo providerInfo) {
                 super.onPostExecute(providerInfo);
 
+                if (canceled) {
```

Comment:
Actually I have a question here. In master branch a variable `canceled` is used to avoid unexpected things after activity is paused, stopped and so on, so I followed it here. But why not use `AsyncTask.cancel` to cancel the task directly here?